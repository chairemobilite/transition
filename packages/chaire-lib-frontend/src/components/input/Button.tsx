/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React, { MouseEvent } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconProp } from '@fortawesome/fontawesome-svg-core';

export type ButtonProps = {
    onClick?: React.MouseEventHandler;
    onKeyUp?: React.KeyboardEventHandler;
    href?: string;
    download?: string; // see: https://stackoverflow.com/questions/10049259/change-name-of-download-in-javascript
    label: string;
    type: 'href' | 'button' | 'submit' | 'reset' | undefined;
    inputRef?: React.RefObject<HTMLButtonElement>;
    isVisible?: boolean;
    align?: string;
    color?: string;
    size?: string;
    icon?: IconProp;
    iconClass?: string;
    iconPath?: string;
    disabled?: boolean;
    title?: string;
    name?: string;
    style?: { [key: string]: string };
};

type AnchorAttributesProps = {
    href: string;
    download?: string; // see: https://stackoverflow.com/questions/10049259/change-name-of-download-in-javascript
    className: string;
};

export class Button extends React.Component<ButtonProps> {
    static defaultProps: Partial<ButtonProps> = {
        isVisible: true,
        align: 'center',
        type: 'button',
        color: 'green',
        size: 'large',
        iconClass: '_icon',
        style: {},
        title: undefined,
        name: undefined,
        disabled: false
    };

    constructor(props: ButtonProps) {
        super(props);
        this.onClickButton = this.onClickButton.bind(this);
    }

    onClickButton(event: MouseEvent): void {
        event.preventDefault();

        if (typeof this.props.onClick === 'function') {
            this.props.onClick(event);
        }
    }

    render(): React.ReactNode {
        if (this.props.isVisible === false) {
            return null;
        }

        if (this.props.type === 'href' && this.props.href) {
            const attributes: AnchorAttributesProps = {
                href: this.props.href,
                className: `button ${this.props.color} ${this.props.size}`
            };

            if (this.props.download) {
                attributes.download = this.props.download;
            }

            return (
                <div className={this.props.align} style={this.props.style || {}}>
                    <a {...attributes}>
                        {this.props.icon && (
                            <FontAwesomeIcon icon={this.props.icon} className={`${this.props.iconClass}`} />
                        )}
                        {this.props.iconPath && (
                            <img
                                className={`_icon ${this.props.iconClass}`}
                                src={this.props.iconPath}
                                alt={this.props.label}
                            />
                        )}{' '}
                        {this.props.label}
                    </a>
                </div>
            );
        } else if (this.props.type !== 'href') {
            const attributes: { [key: string]: any } = {};
            if (this.props.disabled) {
                attributes.disabled = true;
            }
            if (this.props.title) {
                attributes.title = this.props.title;
            }
            if (this.props.name) {
                attributes.name = this.props.name;
            }
            if (this.props.onKeyUp) {
                attributes.onKeyUp = this.props.onKeyUp;
            }
            return (
                <div className={this.props.align} style={this.props.style || {}}>
                    <button
                        {...attributes}
                        type={this.props.type}
                        ref={this.props.inputRef}
                        className={`button ${this.props.color} ${this.props.size}${
                            this.props.disabled ? ' disabled' : ''
                        }`}
                        onClick={this.onClickButton}
                    >
                        {this.props.icon && (
                            <FontAwesomeIcon icon={this.props.icon} className={`${this.props.iconClass}`} />
                        )}
                        {this.props.iconPath && (
                            <img
                                className={`_icon ${this.props.iconClass}`}
                                src={this.props.iconPath}
                                alt={this.props.label}
                            />
                        )}{' '}
                        {this.props.label}
                    </button>
                </div>
            );
        }
    }
}

export default Button;
