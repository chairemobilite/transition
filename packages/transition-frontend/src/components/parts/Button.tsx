/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';

import ButtonCell, { ButtonCellWithConfirm } from './ButtonCell';

interface ButtonProps extends WithTranslation {
    isSelected: boolean;
    children: React.ReactNode;
    key: string;
    /**
     * Whether to flush align to the right the first edit/duplicate/delete
     * action cell. This should be true if no other cell has been flushed to the
     * right. Default: false
     */
    flushActionButtons?: boolean;
    /**
     * If the object can be selected, an edit icon will be shown in the Button.
     * This handler will also be the default handler when the button is clicked
     * if there are no other
     */
    onSelect?: { handler: React.MouseEventHandler; altText?: string };
    /**
     * If the object can be delete, a delete icon will be shown in the Button.
     * It will show a confirmation modal with the message and the delete handler
     * will be called only if the delete is confirmed.
     */
    onDelete?: { handler: React.MouseEventHandler; altText?: string; message: string };
    /**
     * If the object can be duplicated, a duplication icon will be shown in the
     * Button.
     */
    onDuplicate?: { handler: React.MouseEventHandler; altText: string };
}

/**
 * A list element <li> formatted as a Button. Typically to be used in lists of
 * object to be represented with cell-like elements. Children can be added to
 * the props, they will be displayed at the beginning of the button and
 * additional buttons will be added to edit, duplicate and delete the
 * represented object, depending on the props passed.
 * @param props
 * @returns
 */
const Button: React.FunctionComponent<ButtonProps> = (props: ButtonProps) => {
    const deleteText = props.onDelete ? props.onDelete.altText || props.t('main:Delete') : undefined;
    const flushActionButtons = props.flushActionButtons || false;

    return (
        <React.Fragment key={props.key}>
            <li
                className={`_list${props.isSelected === true ? ' _active' : ''}`}
                onClick={props.onSelect ? props.onSelect.handler : undefined}
                key="containerLink"
            >
                {props.children}
                {props.onSelect && (
                    <ButtonCell
                        key={`${props.key}edit`}
                        alignment={flushActionButtons ? 'flush' : 'right'}
                        title={props.onSelect.altText || props.t('main:Edit')}
                    >
                        <img
                            className="_icon-alone"
                            src={'/dist/images/icons/interface/edit_white.svg'}
                            alt={props.onSelect.altText || props.t('main:Edit')}
                        />
                    </ButtonCell>
                )}
                {props.onDuplicate && (
                    <ButtonCell
                        key={`${props.key}duplicate`}
                        alignment={flushActionButtons && !props.onSelect ? 'flush' : 'right'}
                        onClick={props.onDuplicate.handler}
                        title={props.onDuplicate.altText}
                    >
                        <img
                            className="_icon"
                            src={'/dist/images/icons/interface/copy_white.svg'}
                            alt={props.onDuplicate.altText}
                        />
                    </ButtonCell>
                )}
                {props.onDelete && (
                    <ButtonCellWithConfirm
                        key={`${props.key}delete`}
                        alignment={flushActionButtons && !props.onSelect && !props.onDuplicate ? 'flush' : 'right'}
                        onClick={props.onDelete.handler}
                        title={deleteText}
                        confirmButtonColor="red"
                        confirmButtonText={props.onDelete.altText || props.t('main:Delete')}
                        message={props.onDelete.message}
                    >
                        <img
                            className="_icon-alone"
                            src={'/dist/images/icons/interface/delete_white.svg'}
                            alt={deleteText}
                        />
                    </ButtonCellWithConfirm>
                )}
            </li>
            <li className="_clear" key="clearer"></li>
        </React.Fragment>
    );
};

export default withTranslation('main')(Button);
