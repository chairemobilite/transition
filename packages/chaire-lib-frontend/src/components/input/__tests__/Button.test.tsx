/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { Button } from '../Button';
import { render, fireEvent } from '@testing-library/react';
import { faCoffee } from '@fortawesome/free-solid-svg-icons';

const mockOnClick = jest.fn();
const testLabel = 'Test';

test('Default props', () => {
    const { container } = render(<Button
        onClick={mockOnClick}
        label={testLabel}
    />);
    expect(container).toMatchSnapshot();
});

test('Test overriding props with defaults', () => {
    const { container } = render(<Button
        onClick={mockOnClick}
        label={testLabel}
        align="left"
        type="submit"
        color="red"
        size="small"
    />);
    expect(container).toMatchSnapshot();
});

test('Test disabled', () => {
    const { container } = render(<Button
        onClick={mockOnClick}
        label={testLabel}
        align="left"
        type="submit"
        color="red"
        size="small"
        disabled={true}
    />);
    expect(container).toMatchSnapshot();
});

test('Test on click', () => {
    mockOnClick.mockClear();
    const { getByText } = render(<Button
        onClick={mockOnClick}
        label={testLabel}
    />);
    const element = getByText(testLabel);
    fireEvent.click(element);
    expect(mockOnClick).toHaveBeenCalledTimes(1);
});

test('Test invisible', () => {
    const { queryByText } = render(<Button
        onClick={mockOnClick}
        label={testLabel}
        isVisible={false}
    />);
    const element = queryByText(testLabel);
    expect(element).toBeFalsy();
});

test('Test icon path', () => {
    const { container } = render(<Button
        onClick={mockOnClick}
        label={testLabel}
        iconPath="path/to/my/icon.png"
        iconClass="myIconClass"
    />);
    expect(container).toMatchSnapshot();
});

test('Test icon', () => {
    const { container } = render(<Button
        onClick={mockOnClick}
        label={testLabel}
        icon={faCoffee}
        iconClass="myIconClass"
    />);
    expect(container).toMatchSnapshot();
});

