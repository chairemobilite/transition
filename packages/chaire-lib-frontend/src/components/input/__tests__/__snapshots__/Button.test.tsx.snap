// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default props 1`] = `
<div>
  <div
    class="center"
  >
    <button
      class="button green large"
      type="button"
    >
       
      Test
    </button>
  </div>
</div>
`;

exports[`Test disabled 1`] = `
<div>
  <div
    class="left"
  >
    <button
      class="button red small disabled"
      disabled=""
      type="submit"
    >
       
      Test
    </button>
  </div>
</div>
`;

exports[`Test icon 1`] = `
<div>
  <div
    class="center"
  >
    <button
      class="button green large"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-mug-saucer myIconClass"
        data-icon="mug-saucer"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 64c0-17.7 14.3-32 32-32l320 0 64 0c70.7 0 128 57.3 128 128s-57.3 128-128 128l-32 0c0 53-43 96-96 96l-192 0c-53 0-96-43-96-96L96 64zM480 224l32 0c35.3 0 64-28.7 64-64s-28.7-64-64-64l-32 0 0 128zM32 416l512 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 480c-17.7 0-32-14.3-32-32s14.3-32 32-32z"
          fill="currentColor"
        />
      </svg>
       
      Test
    </button>
  </div>
</div>
`;

exports[`Test icon path 1`] = `
<div>
  <div
    class="center"
  >
    <button
      class="button green large"
      type="button"
    >
      <img
        alt="Test"
        class="_icon myIconClass"
        src="path/to/my/icon.png"
      />
       
      Test
    </button>
  </div>
</div>
`;

exports[`Test overriding props with defaults 1`] = `
<div>
  <div
    class="left"
  >
    <button
      class="button red small"
      type="submit"
    >
       
      Test
    </button>
  </div>
</div>
`;
